Parser/Parser.cs

```diff
@@ -84,9 +122,61 @@ NBlock Block ()
    NWriteStmt WriteStmt () 
       => new (Prev.Kind == WRITELN, ArgList ());
 
+   //read-stmt = "read" "(" identlist ")" .
+   NReadStmt ReadStmt ()
+      => new (Prev, IdentList ());
+
+
    // assign-stmt = IDENT ":=" expr .
    NAssignStmt AssignStmt () 
       => new (PrevPrev, Expression ());
+
+   // call-stmt = IDENT arglist .
+   NCallStmt CallStmt ()
+      => new (Prev, ArgList ());
+
+// if-stmt = "if" expression "then" statement["else" statement] .
+   NIfStmt IfStmt () {
+      Expect (IF);
+      var expr = Expression();
+      Expect (THEN);
+      List<NStmt> stmt = new () {Stmt () };
+      while (Match (ELSE)) stmt.Add(Stmt());
+      return new (expr, stmt.ToArray());
+   }
+
+   // while-stmt = "while" expression "do" statement.
+   NWhileStmt WhilefStmt () {
+      Expect (WHILE);
+      var expr = Expression ();
+      Expect (DO);
+      var stmt = Stmt ();
+      return new (expr, stmt);
+   }
+
+   //repeat-stmt = "repeat" statement { ";" statement} "until" expression.
+   NRepeatStmt RepeatfStmt () {
+      List<NStmt> stmts = new ();
+      Expect (REPEAT);
+      stmts.Add(Stmt ());
+      if (Match (SEMI)) stmts.Add (Stmt ());
+      Expect (UNTIL);
+      var expr = Expression ();
+      return new (stmts.ToArray(), expr);
+   }
+
+   //for-stmt = "for" IDENT ":=" expression ("to" | "downto") expression "do" statement.
+   NForStmt ForfStmt () {
+      List<NExpr> exprs = new ();
+      Expect (FOR);
+      var name = Expect (IDENT);
+      Expect (ASSIGN); 
+      exprs.Add(Expression ());
+      exprs.Add (Expression ());
+      Expect (DO);
+      var stmt = Stmt ();
+      return new (name, Prev.Kind == TO, exprs.ToArray(), stmt);

```

Comment:
This is broken. By the time you come to the last line, Prev will no longer be either TO or DOWNTO. It will be some random last token. Did you check if this works?